packages/nouns-contracts/contracts/governance/data/NounsDAOData.sol

```diff
@@ -212,24 +214,44 @@ contract NounsDAOData is OwnableUpgradeable {
 
     /**
      * @notice Add a signature supporting a new candidate to be proposed to the DAO using `proposeBySigs`, by emitting an event with the signature data.
-     * @param signer the signer's address.
+     * Only lets the signer account submit their signature, to minimize potential spam.
      * @param sig the signature bytes.
      * @param expirationTimestamp the signature's expiration timestamp.
-     * @param encodedPropHash the keccak256 hash of the candidate version signed; the hash is performed on the output of
+     * @param proposer the proposer account that posted the candidate proposal with the provided slug.
+     * @param slug the slug of the proposal candidate signer signed on.
+     * @param encodedProp the abi encoding of the candidate version signed; should be identical to the output of
      * the `NounsDAOV3Proposals.calcProposalEncodeData` function.
+     * @param reason signer's reason free text.
      */
     function addSignature(
-        address signer,
         bytes memory sig,
         uint256 expirationTimestamp,
         address proposer,
         string memory slug,
-        bytes32 encodedPropHash,
+        bytes memory encodedProp,
         string memory reason
     ) external {
         if (!propCandidates[proposer][keccak256(bytes(slug))]) revert SlugDoesNotExist();
 
-        emit SignatureAdded(msg.sender, signer, sig, expirationTimestamp, proposer, slug, encodedPropHash, reason);
+        bytes32 sigDigest = NounsDAOV3Proposals.sigDigest(
+            NounsDAOV3Proposals.PROPOSAL_TYPEHASH,
+            encodedProp,
+            expirationTimestamp,
+            address(this)
+        );
+
+        if (!SignatureChecker.isValidSignatureNow(msg.sender, sigDigest, sig)) revert InvalidSignature();
```

Comment:
(adding this as a reminder)
in the context of signing from a gnosis safe:
the safe operator will need to wait until the signature tx is included in a block, and only then they can call this function.
we may want to give some clarity around this in the UI